tests/core/p2p-proto/test_sync.py::test_block_gapfill_syncer fails intermittently

https://app.circleci.com/pipelines/github/ethereum/trinity/6649/workflows/800afaa4-3e06-4772-8e73-84aa70d5504d/jobs/249829/steps

```
                    # We need to give the async calls a moment to settle before we can read the updated
                    # chain gaps.
                    await asyncio.sleep(0.1)
>                   assert chain_with_gaps.chaindb.get_chain_gaps() == ((), 1001)
E                   assert ((), 998) == ((), 1001)
E                     At index 1 diff: 998 != 1001
 ```

Probably just need to wait longer before reading the updated chain gaps?

--- a/tests/core/p2p-proto/test_sync.py
+++ b/tests/core/p2p-proto/test_sync.py
@@ -649,7 +649,7 @@ async def test_block_gapfill_syncer(request,
                 syncer.pause()
                 # We need to give async code a moment to settle before we save the progress to
                 # ensure it has stabilized before we save it.
-                await asyncio.sleep(0.01)
+                await asyncio.sleep(0.25)
                 paused_chain_gaps = chain_with_gaps.chaindb.get_chain_gaps()
 
                 # Consider it victory if after 0.5s no new blocks were written to the database
@@ -668,7 +668,7 @@ async def test_block_gapfill_syncer(request,
 
                 # We need to give the async calls a moment to settle before we can read the updated
                 # chain gaps.
-                await asyncio.sleep(0.1)
+                await asyncio.sleep(0.25)
                 assert chain_with_gaps.chaindb.get_chain_gaps() == ((), 1001)
 
 

--- a/tests/core/p2p-proto/test_sync.py
+++ b/tests/core/p2p-proto/test_sync.py
@@ -618,8 +618,10 @@ async def test_block_gapfill_syncer(request,
         syncer = BodyChainGapSyncer(
             LatestTestChain(chaindb_with_block_gaps.db),
             chaindb_with_block_gaps,
-            MockPeerPoolWithConnectedPeers([client_peer], event_bus=event_bus)
+            MockPeerPoolWithConnectedPeers([client_peer], event_bus=event_bus),
         )
+        # In production, this would be the block time but we want our test to pause/resume swiftly
+        syncer._idle_time = 0.01
         server_peer_pool = MockPeerPoolWithConnectedPeers([server_peer], event_bus=event_bus)
 
         async with run_peer_pool_event_server(
@@ -634,6 +636,10 @@ async def test_block_gapfill_syncer(request,
             async with background_asyncio_service(syncer):
                 chain_with_gaps = LatestTestChain(chaindb_with_block_gaps.db)
                 fat_chain = LatestTestChain(chaindb_1000.db)
+
+                # Ensure we can pause/resume immediately and not just after syncing has started
+                syncer.pause()
+                syncer.resume()
 
                 # Sync the first 100 blocks, then check that pausing/resume works
                 await wait_for_block(

--- a/trinity/sync/beam/chain.py
+++ b/trinity/sync/beam/chain.py
@@ -405,7 +405,9 @@ class BodyChainGapSyncer(Service):
     in tandem with other operations that sync the state.
     """
 
+    _idle_time = BLOCK_BACKFILL_IDLE_TIME
     _starting_tip: BlockHeaderAPI = None
+
     logger = get_logger('trinity.sync.beam.chain.BodyChainGapSyncer')
 
     def __init__(self,
@@ -416,12 +418,18 @@ class BodyChainGapSyncer(Service):
         self._db = db
         self._peer_pool = peer_pool
         self._paused = False
+        self._body_syncer: FastChainBodySyncer = None
         self._resumed = asyncio.Event()
 
     async def _setup_for_next_gap(self) -> None:
         gap_start, gap_end = self._get_next_gap()
         start_num = BlockNumber(gap_start - 1)
         _starting_tip = await self._db.coro_get_canonical_block_header_by_number(start_num)
+
+        if self._paused:
+            # If the syncer was paused while we were busy setting it up throw the current setup
+            # away. A new setup will be performed as soon as `resume()` was called again.
+            raise ValidationError("Syncer was paused by the user")
 
         async def _get_launch_header() -> BlockHeaderAPI:
             return _starting_tip
@@ -476,7 +484,10 @@ class BodyChainGapSyncer(Service):
             raise RuntimeError("Invalid action. Can not pause service that is already paused.")
 
         self._paused = True
-        self._body_syncer.get_manager().cancel()
+        if self._body_syncer:
+            # Pausing the syncer is valid at any point that the service is running but that might
+            # hit a point where we are still resolving the gaps and have no body_syncer set up yet.
+            self._body_syncer.get_manager().cancel()
         self.logger.debug2("BodyChainGapSyncer paused")
 
     def resume(self) -> None:
@@ -503,9 +514,9 @@ class BodyChainGapSyncer(Service):
             except ValidationError:
                 self.logger.debug(
                     "There are no gaps in the chain of blocks at this time. Sleeping for %ss",
-                    BLOCK_BACKFILL_IDLE_TIME
+                    self._idle_time
                 )
-                await asyncio.sleep(BLOCK_BACKFILL_IDLE_TIME)
+                await asyncio.sleep(self._idle_time)
             else:
                 await self.manager.run_service(self._body_syncer)
 

